### DISCUSSION TITLE
I'm facing a problem reading power consumption.

### DISCUSSION THREAD

**Discussion by Bharath-VCG** [oshi/oshi#1839] [Q&A] [answered]
I have 3 machines.

- Laptop with Windows 10
- Laptop with RedHat Linux
- Laptop with Windows Server

The last two machines are for testing. The first one is the development machine.

Below are the issues,

1. I'm getting 0 as power consumption when the laptop is fully charged. But I can able to get power consumption when the laptop is discharging on Windows 10.
2. I'm getting 0 on both Windows Server and RedHat Linux all the time.

I want to get the power consumption. Please help me with this. I have been working on this for more than 2 days. Still, I haven't gotten any clue on the issue.

**Comment by dbwiddis** [comment]
OSHI's implementation for power source is only for the battery.  So if it is plugged in/charging, we do not get any information on power usage.  I'm not even sure where that information can be found in Windows.

Linux data comes from `/sys/class/power_supply/`, see https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-class-power

Can you investigate your linux box and see if there is information there that's not properly being read?

**Reply by Bharath-VCG** [reply]
Thanks for your reply.

![image](https://user-images.githubusercontent.com/97423822/148747912-7df16344-b7c9-4487-a26d-9186f9f2674b.png)

I have no file with that name in the directory you mentioned. Any idea why RedHat doesn't have it?

**Reply by dbwiddis** [reply]
You do have many files in that power_supply sysfs path. Not sure which one you say isn't there?  Is there anything under `power`?

`current_now` and `voltage_now` should give you the information you need (Power = current x voltage), but OSHI reads the `uevent` file and looks for relevant keys: `POWER_SUPPLY_POWER_NOW` should be the power usage rate.  Can you paste the contents of that `uevent` file?

**Reply by Bharath-VCG** [reply]
> You do have many files in that power_supply sysfs path. Not sure which one you say isn't there? Is there anything under `power`?
> 
> `current_now` and `voltage_now` should give you the information you need (Power = current x voltage), but OSHI reads the `uevent` file and looks for relevant keys: `POWER_SUPPLY_POWER_NOW` should be the power usage rate. Can you paste the contents of that `uevent` file?

![image](https://user-images.githubusercontent.com/97423822/149942767-5be237d8-9907-44a5-b2a4-f49ba3ac3d01.png)

There is no key with that name.

Sorry for the late response.

**Comment by dbwiddis** [comment, marked as answer]
OK, looks like a bug in OSHI where it is not reading current (which is documented) but is reading voltage (which apparently exists on some distros but is not documented).  We should change the code to default to calculating power by multiplying current and voltage.

Current: 
> Represented in microamps. Negative values are used for discharging batteries, positive values for charging batteries 

1000 microamps = 1 mA.

Voltage:
> Represented in microvolts.

8402000 microvolts = 8.402 V

So power = 1mA x 8.402V = 8.402 mW

**Reply by Bharath-VCG** [reply]
Thanks for your reply.
It helped me a lot.

**Reply by dbwiddis** [reply]
I just released 6.1.0 with this fix.

**Comment by dbwiddis** [comment]
Fixed in #1859
